WordPress/gutenberg
# Lazily load CodeMirror assets

_This PR should have its base set to `master` once https://github.com/WordPress/gutenberg/pull/4348 is merged!_

## 🤠 What this is 

In https://github.com/WordPress/gutenberg/pull/4348, we added the CodeMirror library so as to provide syntax highlighting and a nice editor experience for writing HTML blocks.

Unfortunately, though, this means an extra [~1.9 MB of assets](https://github.com/WordPress/gutenberg/pull/4348#issuecomment-356134596) are loaded when initialising Gutenberg.

The solution is to lazily load these assets. That is, rather than loading the CodeMirror library when Gutenberg initialises, we load it when a HTML block is first inserted into the post.

This PR accomplishes this by changing the `CodeEditor` component to use a new `withLazyDependencies` HOC:

https://github.com/WordPress/gutenberg/blob/87714ec96bed381df3b5e2c8d521d569d8f5305d/components/code-editor/index.js#L110-L129

When `CodeEditor` is mounted, a `<script src="/wp-admin/wp-load-scripts.php?load...">` node and a `<link rel="stylesheet" href="/wp-admin/wp-load-styles.php?load=...">` node is inserted into the DOM which causes the assets to be asynchronously loaded.

Check out [the included README](https://github.com/WordPress/gutenberg/blob/87714ec96bed381df3b5e2c8d521d569d8f5305d/components/higher-order/with-lazy-dependencies/README.md) to read more about `withLazyDependencies `.

### ✅ How to test

1. Insert a HTML block
2. Type some code into it. You should see syntax highlighting
3. Save the post and reload the page. The HTML block with syntax highlighting should remain
